### DISCUSSION TITLE
Trying to parse a URL into a Route enum

### DISCUSSION THREAD

**Discussion by ghost** [pointfreeco/swift-parsing#179] [Q&A] [answered]
Hello :)

I'm trying to parse an URL of the shape
```swift
https://my.website.com/sign-in/mobile?token=SomeL0ngSequence0fChars&flow=sign-in
```
so I tried the following
```swift
enum AppRoute {
    case signIn(token: String)
}

let router = OneOf {
    Route(AppRoute.signIn(token:)) {
        Path(FromUTF8View { "sign-in".utf8 })
        Path(FromUTF8View { "mobile".utf8 })
        Query("token", PrefixUpTo("&"))
        Query("flow", "sign-in")
    }
}
```
where `Route`, `Path`, and `Query` are copied from https://github.com/pointfreeco/swift-parsing/tree/main/Sources/swift-parsing-benchmark/URLRequestRouter. The problem I'm having is that the token query parser is causing the compiler to require `Void` and `Substring` to be equivalent:

> Initializer 'init(_:with:)' requires the types 'Substring' and 'Void' be equivalent

I'm at a loss trying to understand what I'm doing wrong and how to fix it, and would greatly appreciate any help.

Many thanks in advance!

**Comment by stephencelis** [comment, marked as answer]
@wtruppel That's a pretty bad error message 😞

The main issue is that `PrefixUpTo` (and most parsers on strings) will output a `Substring`, so to bundle things up into your enum you must first convert the value back to a string:

```swift
Query("token", PrefixUpTo("&").map(String.init))
```

Please note, though, that query parsers are already split up to work on individual items, so the "&" character shouldn't be relevant at this point. I think you could instead do:

```swift
Query("token", Rest().map(String.init))
```

**Reply by ghost** [reply]
Ah, awesome!! Thanks, @stephencelis ! 💯
